PXB-3283: Column %s precise type mismatch error while importing a tablespace (trunk)

venkatesh-prasad-v opened the pull request:
https://perconadev.atlassian.net/browse/PXB-3283

Problem & Analysis
==================
MySQL always stores temporal fields in my_charset_numeric charset.

When the collation of the table is changed using ALTER TABLE CONVERT TO CHARACTER SET, the collation of the temporal field was also updated in SDI.

When PXB prepares the backup (with --export) having such table, the .cfg file created by PXB sets the prtype based on the collation that is present in the SDI.

While importing the tablespace, the server fails with error

    "Schema mismatch (Column <col_name> precise type mismatch.)"

since InnoDB detects different prtype for the temporal column.

However, this problem is not seen when we import .cfg file generated by FLUSH TABLE FOR EXPORT since FLUSH TABLE FOR EXPORT always uses the charset of the table->field (my_charset_numeric) and sets prtype using my_charset_numeric as the collation.

Solution
========
The problem of server allowing ALTER TABLE CONVERT TO CHARACTER SET command to alter the character set temporal feilds has been fixed in Percona Server by PS-9129. But in order to remain consistent with the server behavior, PXB nor forces all temporal fields to use my_charset_numeric as the charset until MySQL server provides support for a separate charset/collation for temporal fields.

Testing Done
========
Jenkins: https://pxb.cd.percona.com/view/PXB%208.1/job/percona-xtrabackup-8.1-multijob/42/console
_Test runs in progress_

Commit:
PXB-3283: Column %s precise type mismatch error while importing a tablespace

https://perconadev.atlassian.net/browse/PXB-3283

Problem & Analysis
==================
MySQL always stores temporal fields in my_charset_numeric charset.

When the collation of the table is changed using ALTER TABLE
CONVERT TO CHARACTER SET, the collation of the temporal field was
also updated in SDI.

When PXB prepares the backup (with --export) having such table, the .cfg
file created by PXB sets the prtype based on the collation that is
present in the SDI.

While importing the tablespace, the server fails with error

    "Schema mismatch (Column <col_name> precise type mismatch.)"

since InnoDB detects different prtype for the temporal column.

However, this problem is not seen when we import .cfg file generated by
FLUSH TABLE FOR EXPORT since FLUSH TABLE FOR EXPORT always uses the
charset of the table->field (my_charset_numeric) and sets prtype using
my_charset_numeric as the collation.

Solution
========
The problem of server allowing ALTER TABLE CONVERT TO CHARACTER SET
command to alter the character set temporal feilds has been fixed in
Percona Server by PS-9129. But in order to remain consistent with the
server behavior, PXB now forces all temporal fields to use
my_charset_numeric as the charset until MySQL server provides support
for a separate charset/collation for temporal fields.

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/not_signed)](https://cla-assistant.percona.com:443/percona/percona-xtrabackup?pullRequest=1560) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/percona-xtrabackup?pullRequest=1560) before we can accept your contribution.<br/><hr/>**Venkatesh Prasad** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/percona-xtrabackup?pullRequest=1560) it.</sub>

Commit:
PXB-3283: Column %s precise type mismatch error while importing a
tablespace (trunk)

https://perconadev.atlassian.net/browse/PXB-3283

Merge branch 'PXB-3283-8.0-gca' into PXB-3283-trunk

Problem & Analysis
==================
MySQL always stores temporal fields in my_charset_numeric charset.

When the collation of the table is changed using ALTER TABLE
CONVERT TO CHARACTER SET, the collation of the temporal field was
also updated in SDI.

When PXB prepares the backup (with --export) having such table, the .cfg
file created by PXB sets the prtype based on the collation that is
present in the SDI.

While importing the tablespace, the server fails with error

    "Schema mismatch (Column <col_name> precise type mismatch.)"

since InnoDB detects different prtype for the temporal column.

However, this problem is not seen when we import .cfg file generated by
FLUSH TABLE FOR EXPORT since FLUSH TABLE FOR EXPORT always uses the
charset of the table->field (my_charset_numeric) and sets prtype using
my_charset_numeric as the collation.

Solution
========
The problem of server allowing ALTER TABLE CONVERT TO CHARACTER SET
command to alter the character set temporal feilds has been fixed in
Percona Server by PS-9129. But in order to remain consistent with the
server behavior, PXB now forces all temporal fields to use
my_charset_numeric as the charset until MySQL server provides support
for a separate charset/collation for temporal fields.
